activation/nipost.go

```diff
@@ -211,21 +174,28 @@ func (nb *NIPostBuilder) BuildNIPost(ctx context.Context, challenge *types.NIPos
 
 		submitCtx, cancel := context.WithDeadline(ctx, poetRoundStart)
 		defer cancel()
-		poetRequests, err := nb.submitPoetChallenges(submitCtx, poetProofDeadline, challengeHash.Bytes())
-		if err != nil {
+
+		if err := nb.submitPoetChallenges(submitCtx, poetProofDeadline, challenge.Hash().Bytes()); err != nil {
 			return nil, fmt.Errorf("submitting to poets: %w", err)
 		}
-		if len(poetRequests) == 0 {
-			return nil, &PoetSvcUnstableError{msg: "failed to submit challenge to any PoET", source: submitCtx.Err()}
+		count, err := nipost.PoetRegistrationCount(nb.localDB, nb.signer.NodeID())
+		if err != nil {
+			// TODO(mafa): panic instead?
+			return nil, &PoetSvcUnstableError{msg: "failed to submit challenge to any PoET", source: err}
+		}
+		if count == 0 {
+			return nil, &PoetSvcUnstableError{msg: "failed to submit challenge to any PoET"}
 		}
+	}
 
-		nb.state.Challenge = challengeHash
-		nb.state.PoetRequests = poetRequests
-		nb.persistState()
+	poetProofRef, err := nipost.PoetProofRef(nb.localDB, nb.signer.NodeID())
+	if err != nil {
+		// TODO(mafa): panic instead?
```

Comment:
same as above; before if the state could not be read from disk it would just assume that there were not registrations yet.